# Chapter 2 - The Basic Theory of Interest

## Exercise 1.

**(a)** Suppose we invested \\$1 in 1776 at 3.3% interest compounded yearly and we assume the current date to be 2009 (the date of publication), then we have for $r = 0.033$ and $n = (2009 - 1) - (1776 + 1) = 231$ years. This calculation allows for the yearly interest at the start of 2009 and accounts for the fact that interest only begins to accrue from 1777. Thus we have the present value of the investment given by

$$
\begin{equation*}
    \begin{split}
        \text{PV} &= {(1 + r)}^n \\
                  &= {1.033}^{231} \\
                  &\approx \$1,807.90.
    \end{split}
\end{equation*}
$$

We use the following to calculate the approximate value:

In [1]:
pv = 1.033**231
print(pv)

1807.899443302222


The closest estimate is thus \\$1,000.

**(b)** If the interest rate were 6.6% compounded annually, we would have the following present value

$$
\begin{equation*}
    \begin{split}
        \text{PV} &= {(1 + r)}^n \\
                  &= {1.066}^{231} \\
                  &\approx \$2,581,750.58.
    \end{split}
\end{equation*}
$$

In [2]:
pv = 1.066**231
print(pv)

2581750.581539662


Thus the new approximation would be \\$2,581,750.58.

## Exercise 2.

We have 

$$
\begin{equation*}
    \begin{split}
                             {(1 + r)}^n &= 2 \\
        \Leftrightarrow \ln{{(1 + r)}^n} &= \ln{2} \\ 
           \Leftrightarrow n\ln{(1 + r)} &= \ln{2} \\
                       \Leftrightarrow n &= \frac{\ln{2}}{\ln{(1 + r)}}.
    \end{split}
\end{equation*}
$$

Then, applying the approximation $ln{(1 + r)} \approx r$ for small $r$, we have

$$
\begin{equation*}
    \begin{split}
        n &\approx \frac{\ln{2}}{r} \\
          &= \frac{.69}{r}.
    \end{split}
\end{equation*}
$$

Thus, using $i = 100r$, we have

$$
\begin{equation*}
    \begin{split}
        n &\approx \frac{\ln{2}}{r} \\
          &= \frac{.69}{(i / 100)} \\
          &= \frac{69}{i}.
    \end{split}
\end{equation*}
$$

Now, using the better approximation $\ln{(1 + r)} \approx r - \frac{1}{2}r^2$, we have for $r \approx .08$ we conclude that

$$
\begin{equation*}
    \begin{split}
        n &\approx \frac{\ln{2}}{r - \frac{1}{2}r^2} \\
          &= \frac{.69}{.08 - \frac{1}{2}{.08}^2}.
    \end{split}
\end{equation*}
$$

Using the following calculation

In [3]:
n = .69 / (.08 - (.5 * (.08**2)))
print(n)

8.984374999999998


For $72 / i = 72 / 100r \approx 72 / (100 \cdot .08)$, we have the following calculation

In [4]:
n_approx = 72 / (100 * .08)
print(n_approx)
print(abs(n_approx - n))


9.0
0.015625000000001776


Thus we can see that $n \approx 72 / i$.


## Exercise 3.

**(a)** An annual rate of 3% compounded monthly will produce an increase of ${(1 + r)}^n = {(1.0025)}^{12}$, which we approximate by the following

In [5]:
increase = (1.0025 ** 12)
print(increase)

1.0304159569135067


Thus the effective rate is given by 3.04%.

**(b)** An annual rate of 18% compounded monthly will produce an increase of ${(1.015)}^{12}$, which we approximate by the following

In [6]:
increase = (1.015 ** 12)
print(increase)

1.195618171461534


Thus the effective rate is given by 19.56%.

**(c)** An annual rate of 18% compounded quarterly will produce an increase of ${(1.045)}^4$, which we approximate by the following

In [7]:
increase = (1.045 ** 4)
print(increase)

1.1925186006249997


Thus the effective rate is given by 19.25%.

## Exercise 4.

We have the following definition of Newton's method and the appropriate functions:

In [8]:
from typing import Callable


def newtons_method(f: Callable[[float], float], f_p: Callable[[float], float], lambda_k: float) -> float:
    return lambda_k - (f(lambda_k) / f_p(lambda_k))


lambdas: list[float] = [1.0]


def f(x: float) -> float:
    return -1.0 + x + x**2


def f_p(x: float) -> float:
    return 1.0 + 2.0 * x


for k in range(1, 5):
    lambdas.append(newtons_method(f, f_p, lambdas[k - 1]))


for lambda_bar in lambdas:
    print(f"lambda_bar = {lambda_bar}, f(lambda_bar) = {f(lambda_bar)}")

lambda_bar = 1.0, f(lambda_bar) = 1.0
lambda_bar = 0.6666666666666667, f(lambda_bar) = 0.11111111111111127
lambda_bar = 0.6190476190476191, f(lambda_bar) = 0.0022675736961451642
lambda_bar = 0.6180344478216818, f(lambda_bar) = 1.0265159329780893e-06
lambda_bar = 0.618033988749989, f(lambda_bar) = 2.1066481892262345e-13


Thus we conclude that the approximation after four estimates is $\bar{\lambda} \approx 0.618033988749989$ and $f(\bar{\lambda}) \approx 2.1066481892262345 \times {10}^{-13} \approx 0$.

## Exercise 5.

Given payments $x_k = 500000$, for $k = 0, 1, \ldots, 20$, and $r = 0.1$, we have the following present value: 

$$
\begin{equation*}
    \begin{split}
        \text{PV} &= \sum_{k = 0}^n{\frac{x_k}{{(1 + r)}^k}} \\
                  &= \sum_{k = 0}^{20}{\frac{500000}{1.1^k}} \\
                  &\approx \$4,756,781.86,
    \end{split}
\end{equation*}
$$

so that the present value of the prize is \$4,756,781.86.

We use the following to calculate the approximate present value:

In [9]:
n = 20
r = 0.1
pv = 0.0
x_k = 500000
for k in range(0, n + 1):
    pv += x_k / ((1 + r) ** k)

print(pv)

4756781.85987928


## Exercise 6.

Assuming that the couple wants to stay in the apartment for 6 months, we have two cash flow streams $\mathbf{x} = (1000, 1000, 1000, 1000, 1000, 1000)$ and $\mathbf{z} = (1000, 900, 900, 900, 900, 900)$, with $r = 0.12$, and $k = 0, 1, \ldots, 6 = n$, so that we have the following present values:

$$
\begin{equation*}
    \begin{split}
        {\text{PV}}_{\mathbf{x}} &= \sum_{k = 0}^n{\frac{x_k}{{(1 + r)}^k}} \\
                                 &= \sum_{k = 0}^6{\frac{1000}{1.12^k}} \\
                                 &\approx \$5,111.41,
    \end{split}
\end{equation*}
$$

and

$$
\begin{equation*}
    \begin{split}
        {\text{PV}}_{\mathbf{z}} &= \sum_{k = 0}^n{\frac{z_k}{{(1 + r)}^k}} \\
                                 &= 1000 + \sum_{k = 1}^6{\frac{900}{1.12^k}} \\
                                 &\approx \$4,700.27,
    \end{split}
\end{equation*}
$$

so that the couple should choose to move to the cheaper apartment.

We use the following to calculate the approximate present values:

In [10]:
n = 6
r = 0.12

pv_x = 0.0
xs = [1000] * (n + 1)
for k in range(0, n + 1):
    pv_x += xs[k] / ((1 + r) ** k)

print(pv_x)

pv_z = 0.0
zs = [1000] + ([900] * (n + 1))
for k in range(0, n + 1):
    pv_z += zs[k] / ((1 + r) ** k)

print(pv_z)

5111.407323522325
4700.2665911700915


Assuming that the couple wants to stay for 1 year, we would have $n = 12$, with ${\text{PV}}_{\mathbf{x}} \approx \$7,194.37$ and $\text{PV}_{\mathbf{z}} \approx \$6,574.94$, so that again the couple should choose to move to the cheaper apartment. 

We use the following to calculate the approximate present values:

In [11]:
n = 12
r = 0.12

pv_x = 0.0
xs = [1000] * (n + 1)
for k in range(0, n + 1):
    pv_x += xs[k] / ((1 + r) ** k)

print(pv_x)

pv_z = 0.0
zs = [1000] + ([900] * (n + 1))
for k in range(0, n + 1):
    pv_z += zs[k] / ((1 + r) ** k)

print(pv_z)

7194.374225455087
6574.936802909578


## Exercise 7

Delaying the harvest another year with the same prevailing interest rate of 10% and the return $R$ would yield the formula

$$
    \text{NPV} = -1 + R / {(1.1)}^2 < 1.48,
$$

so that we have

$$
\begin{equation*}
    \begin{split}
            R / {(1.1)}^2 &< 2.48 \\
        \Leftrightarrow R &< {(1.1)}^2 \cdot 2.48 \\
                          &= 2.904,
    \end{split}
\end{equation*}
$$

so that $x = 2.904$.

## Exercise 8

To calculate the IRR on an incremental basis, we subtract each cash flow of the less expensive of the two options considered from the more expensive option, yielding positive cash flows for all but the initial payment. Thus we have the following table of cash flows:

| |Option A vs Option B|Option B vs Option C|Option A vs Option C|
|-|--------|--------|--------|
|$x_0$|-24,000|-5,000|-29,000|
|$x_1 \cdots x_4$|6,000|400|6,400|
|$x_5$|10,000|2,000|12,000|

Thus we have the following polynomial for considering changing from Option A to Option B:

$$
    -24,000 + 6,000c + 6,000c^2 + 6,000c^3 + 6,000c^4 + 10,000c^5,
$$

or, equivalently


$$
    p(r) = -24,000 + \frac{6,000}{1 + r} + \frac{6,000}{{(1 + r)}^2} + \frac{6,000}{{(1 + r)}^3}
    + \frac{6,000}{{(1 + r)}^4} + \frac{10,000}{{(1 + r)}^5}.
$$

Using Newton's method from Exercise 4, we have the following calculation:

In [12]:
def cash_flows(r: float) -> float:
    return (
        -24000
        + 6000 / (1 + r)
        + 6000 / (1 + r) ** 2
        + 6000 / (1 + r) ** 3
        + 6000 / (1 + r) ** 4
        + 10000 / (1 + r) ** 5
    )


def cash_flows_p(r: float) -> float:
    return (
        -6000 / (1 + r) ** 2
        - 12000 / (1 + r) ** 3
        - 18000 / (1 + r) ** 4
        - 24000 / (1 + r) ** 5
        - 50000 / (1 + r) ** 6
    )

rs = [0.1]
for k in range(1, 6):
    rs.append(newtons_method(cash_flows, cash_flows_p, rs[k - 1]))

for r in rs:
    print(f"r = {r}, f(r) = {cash_flows(r)}")

r = 0.1, f(r) = 1228.4059086873058
r = 0.11770177473592074, f(r) = 45.50574259475525
r = 0.11840909697075933, f(r) = 0.06799356632109266
r = 0.11841015700002991, f(r) = 1.5231853467412293e-07
r = 0.11841015700240459, f(r) = 9.094947017729282e-12
r = 0.11841015700240473, f(r) = -7.275957614183426e-12


Thus we have $r_0 \approx 0.118$, i.e., an IRR of approximately 11.8%, indicating that a change from A to B is justified on the basis of the IRR.

# Exercise 9

Since the new roof lasts 20 years with a cost of \\$20,000, we first calculate how much the new roof would cost, if we assume that we can continue replacing it every 20 years indefinitely. Thus we have the present value given by

$$
\begin{equation*}
    \begin{split}
        \text{PV}_0 &= \frac{C}{1 - {(1 + r)}^{-n}} \\
                    &= \frac{\$20,000}{{1 - (1 + 0.05)}^{-20}} \\
                    &\approx \$32,097.03,
    \end{split}
\end{equation*}
$$

which we calculate as follows:

In [13]:
pv_0 = 20000 / (1 - (1.05**(-20)))
print(pv_0)

32097.034876276513


Since the old roof can be replaced after 5 years, we should calculate the present value as the present value of the new roof minus five years using the value $\text{PV}_0$.

$$
\begin{equation*}
    \begin{split}
        \text{PV}_5 &= \frac{\text{PV}_0}{{(1 + r)}^n} \\
                    &\approx \frac{\$32,097.35}{{(1 + 0.05)}^{5}} \\
                    &\approx \$25,148.87,
    \end{split}
\end{equation*}
$$

which we calculate as follows:

In [14]:
pv_5 = pv_0 / (1.05**5)
print(pv_5)

25148.86669161336


Finally, we have the value of the new roof given by the value the new roof purchased immediately less that of purchasing it in 5 years, and then continuing to purchase a new roof indefinitely. Thus we have

$$
\begin{equation*}
    \begin{split}
        \text{PV} &= \text{PV}_0 - \text{PV}_5 \\
                  &\approx \$32,097.35 - \$25,148.87 \\
                  &\approx \$6,948.17, 
    \end{split}
\end{equation*}
$$

from the following:

In [15]:
print(pv_0 - pv_5)

6948.1681846631545


## Exercise 10

**(a)** We use the following calculations to complete the table as follows:

In [16]:
from copy import deepcopy

opt1 = {
    "barrels_produced": [80_000.0, 70_000.0, 50_000.0, 30_000.0, 10_000.0],
    "gross_revenue": [1_600_000.0, 1_400_000.0, 1_000_000.0, 600_000.0, 200_000.0],
    "net_income": [1_200_000.0, 1_000_000.0, 500_000.0, 200_000.0, 50_000.0],
    "depletion_allowance": [],
    "taxable_income": [],
    "tax": [],
    "after_tax_income": [],
}
opt2 = deepcopy(opt1)

for i in range(len(opt1["barrels_produced"])):
    # Option 1
    depletion_of_gross = 0.22 * opt1["gross_revenue"][i]
    depletion_of_net = opt1["net_income"][i] * 0.5
    opt1["depletion_allowance"].append(
        depletion_of_gross
        if depletion_of_gross < depletion_of_net
        else depletion_of_net
    )
    opt1["taxable_income"].append(
        opt1["net_income"][i] - opt1["depletion_allowance"][i]
    )
    opt1["tax"].append(opt1["taxable_income"][i] * 0.45)
    opt1["after_tax_income"].append(opt1["net_income"][i] - opt1["tax"][i])

    # Option 2
    opt2["depletion_allowance"].append(opt2["barrels_produced"][i] * 5)
    opt2["taxable_income"].append(
        opt2["net_income"][i] - opt2["depletion_allowance"][i]
    )
    opt2["tax"].append(opt2["taxable_income"][i] * 0.45)
    opt2["after_tax_income"].append(opt2["net_income"][i] - opt2["tax"][i])

print(opt1)
print(opt2)

{'barrels_produced': [80000.0, 70000.0, 50000.0, 30000.0, 10000.0], 'gross_revenue': [1600000.0, 1400000.0, 1000000.0, 600000.0, 200000.0], 'net_income': [1200000.0, 1000000.0, 500000.0, 200000.0, 50000.0], 'depletion_allowance': [352000.0, 308000.0, 220000.0, 100000.0, 25000.0], 'taxable_income': [848000.0, 692000.0, 280000.0, 100000.0, 25000.0], 'tax': [381600.0, 311400.0, 126000.0, 45000.0, 11250.0], 'after_tax_income': [818400.0, 688600.0, 374000.0, 155000.0, 38750.0]}
{'barrels_produced': [80000.0, 70000.0, 50000.0, 30000.0, 10000.0], 'gross_revenue': [1600000.0, 1400000.0, 1000000.0, 600000.0, 200000.0], 'net_income': [1200000.0, 1000000.0, 500000.0, 200000.0, 50000.0], 'depletion_allowance': [400000.0, 350000.0, 250000.0, 150000.0, 50000.0], 'taxable_income': [800000.0, 650000.0, 250000.0, 50000.0, 0.0], 'tax': [360000.0, 292500.0, 112500.0, 22500.0, 0.0], 'after_tax_income': [840000.0, 707500.0, 387500.0, 177500.0, 50000.0]}


Thus we have the following table for Option 1:

| **Year** | **Barrels produced** | **Gross revenue (\$)** | **Net income (\$)** | **Depletion allowance (\$)** | **Taxable income (\$)** | **Tax(\$)** | **After-tax income (\$)** |
|---------:|---------------------:|------------------:|---------------:|------------------------:|-------------------:|--------:|---------------------:|
|1|80,000|1,600,000|1,200,000|352,000|848,000|381,600|818,400|
|2|70,000|1,400,000|1,000,000|308,000|692,000|311,400|688,600|
|3|50,000|1,000,000|500,000|220,000|280,000|126,000|374,000|
|4|30,000|600,000|200,000|100,000|100,000|45,000|155,000|
|5|10,000|200,000|50,000|25,000|25,000|11,250|38,750|

The following table is for Option 2:

| **Year** | **Barrels produced** | **Gross revenue (\$)** | **Net income (\$)** | **Depletion allowance (\$)** | **Taxable income (\$)** | **Tax(\$)** | **After-tax income (\$)** |
|---------:|---------------------:|------------------:|---------------:|------------------------:|-------------------:|--------:|---------------------:|
|1|80,000|1,600,000|1,200,000|400,000|800,000|360,000|840,000|
|2|70,000|1,400,000|1,000,000|350,000|650,000|292,500|707,500|
|3|50,000|1,000,000|500,000|250,000|250,000|112,500|387,500|
|4|30,000|600,000|200,000|150,000|50,000|22,500|177,500|
|5|10,000|200,000|50,000|50,000|0|0|50,000|

**(b)** We are given an interest rate of 20\%, i.e., $r = 0.2$, and the two cash flows $(x_0, x_1, \ldots, x_5)$ for Option 1 and $(y_0, y_1, \ldots y_5)$ for Option 2, where $x_0 = y_0 = -\$1,000,000$ is the initial investment made at time $0$, and $x_1, \ldots, x_5$ (resp. $y_1, \ldots, y_5$) are the after-tax incomes of years 1 through 5. Then, we have the $\text{PV}_1$ for Option 1 given by

$$
\begin{equation*}
    \begin{split}
        \text{PV}_1 &= \sum_{k = 0}^n{\frac{x_k}{(1 + r)^k}} \\ 
                    &= \sum_{k = 0}^5{\frac{x_k}{1.2^k}} \\ 
                    &\approx \$466,951.61,
    \end{split}
\end{equation*}
$$

which we calculate as follows:

In [17]:
xs = [-1e6] + opt1["after_tax_income"]
pv_1 = sum(xs[k] / 1.2**k for k in range(len(xs)))
print(pv_1)

466951.61394032923



Likewise, we have $\text{PV}_2$ for Option 2 given by

$$
\begin{equation*}
    \begin{split}
        \text{PV}_2 &= \sum_{k = 0}^n{\frac{y_k}{(1 + r)^k}} \\ 
                    &= \sum_{k = 0}^5{\frac{y_k}{1.2^k}} \\ 
                    &\approx \$521,260.93,
    \end{split}
\end{equation*}
$$

which we calculate as follows:

In [18]:
ys = [-1e6] + opt2["after_tax_income"]
pv_2 = sum(ys[k] / 1.2**k for k in range(len(ys)))
print(pv_2)

521260.9310699589


Now, for the internal rate of return, we have from the main theorem of internal rate of return that there is a unique positive root to the equation

$$
    0 = x_0 + x_1c + x_2c^2 + \cdots + x_5 c^5,
$$

where $\sum_{k=0}^5 x_k > 0$, and likewise for $(y_0, y_1, \ldots, y_5)$. Thus, we again use Newton's method to calculate the IRR for Option 1 as follows:

In [19]:
xrs = [0.1]

def cash_flow_xs(r: float) -> float:
    return sum([xs[k] / (1 + r)**k for k in range(len(xs))])

def cash_flow_xs_p(r: float) -> float:
    return sum([-k * xs[k] / (1 + r)**(k + 1) for k in range(1, len(xs))])

for k in range(1, 7):
    xrs.append(newtons_method(cash_flow_xs, cash_flow_xs_p, xrs[k - 1]))

for r in xrs:
    print(f"r = {r}, f(r) = {cash_flow_xs(r)}")

r = 0.1, f(r) = 724010.4314782271
r = 0.34363071374584314, f(r) = 201105.8185157823
r = 0.4742337486516006, f(r) = 27078.24616950297
r = 0.49774257858744564, f(r) = 650.4043175782472
r = 0.4983353999034056, f(r) = 0.39462663739323034
r = 0.4983357600289833, f(r) = 1.453181539545767e-07
r = 0.4983357600291159, f(r) = 8.003553375601768e-11


Thus the IRR for Option 1 is given by approximately 49.83\%. For Option 2, we have the following:

In [20]:
yrs = [0.1]

def cash_flow_ys(r: float) -> float:
    return sum([ys[k] / (1 + r)**k for k in range(len(ys))])

def cash_flow_ys_p(r: float) -> float:
    return sum([-k * ys[k] / (1 + r)**(k + 1) for k in range(1, len(ys))])

for k in range(1, 7):
    yrs.append(newtons_method(cash_flow_ys, cash_flow_ys_p, yrs[k - 1]))

for r in yrs:
    print(f"r = {r}, f(r) = {cash_flow_ys(r)}")

r = 0.1, f(r) = 791762.5472676354
r = 0.3526927755360726, f(r) = 228255.65886526692
r = 0.4975663208883423, f(r) = 33681.03905641089
r = 0.5270231574661897, f(r) = 997.096263199669
r = 0.5279492576632379, f(r) = 0.9297593904293535
r = 0.5279501228339057, f(r) = 8.099632395897061e-07
r = 0.5279501228346595, f(r) = -2.091837814077735e-11


Thus the IRR for Option 2 is given by approximately 52.80\%.

Since Option 2 has both the greater present value at 20\% and the greater internal rate of return, Option 2 is preferred under either criterion.